### PR TITLE
Forced load imbalance

### PR DESCRIPTION
**Pull request by nanz-nv**
# What does this PR do ?
This PR tries to add a utility to force load imbalance among experts.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -3335,6 +3335,11 @@ def _add_moe_args(parser):
                        'The default value 1e-3 is same as that used in DeepSeekV3.')
     group.add_argument('--moe-router-force-load-balancing', action='store_true',
                        help='[Experimental] Force override routing to balance token distribution using random logits for MoE routers, supporting naive top-k and group-limited top-k. This experimental feature is for benchmarking purposes only!')
+    group.add_argument('--moe-router-force-biased', type=float, default=None,
```
Comment (Victarry) [inline megatron/training/arguments.py:3335]: Overall LGTM. I think the argument name and help message could be refined to make it more clear to users.

Suggested changes:

```
    group.add_argument('--moe-router-force-expert-bias-std', type=float, default=None,
                       help='[Experimental] Apply random expert bias in normal distribution with specified std to router logits. The random seeds are shared  across all ranks.'
                       'If the specified value is positive, generates new random bias each forward pass. '
                       'If the specified value is negative, generates bias once per layer and reuses it (abs value is std).'
                       'This experimental feature is for benchmarking purposes only!')
```

Comment (nanz-nv) [reply megatron/training/arguments.py:3335]: Thanks. The help message has been updated.

**Comment by Skylion007** [comment]
Thank you, this is very useful for MLSys debugging.

**Comment by yaox12** [comment]
The lint issue needs to be fixed.

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -616,6 +616,13 @@ class TransformerConfig(ModelParallelConfig):
     """[Experimental] Force load balancing with random logits for MoE router, supports naive topk 
     and group-limited topk. This is an experimental feature and only for benchmark."""
 
+    moe_router_force_biased: Optional[float] = None
```
Comment (Skylion007) [inline megatron/core/transformer/transformer_config.py:619]: One API issue is that can the be fused together as an enum string parsing with the current Boolean value for force load balancing? Or at least have some though going into other force routing strategies for benchmarking, for instance, you may want to try the worst case of all experts being biased to the same GPU group to test worst case memory scenarios.

**Comment by yanring** [comment]
Communicated with Nan. Nan will find some time to file the main PR in the near future.
